engine/server/engine/centralized_logs/client_implementations/persistent_volume/stream_logs_strategy/per_week_stream_logs_strategy.go

```diff
@@ -198,39 +146,141 @@ func (strategy *PerWeekStreamLogsStrategy) getRetainedLogsFilePaths(filesystem v
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "An error occurred opening the logs file at the following path: %v", pathStr)
+		}
+		fileReaders = append(fileReaders, logsFile)
+	}
+
+	// combine log file readers into a single reader
+	combinedLogsReader := io.MultiReader(fileReaders...)
+
+	return bufio.NewReader(combinedLogsReader), nil
+}
+
+func (strategy *PerWeekStreamLogsStrategy) streamAllLogs(
+	ctx context.Context,
+	logsReader *bufio.Reader,
 	logsByKurtosisUserServiceUuidChan chan map[service.ServiceUUID][]logline.LogLine,
 	serviceUuid service.ServiceUUID,
-	conjunctiveLogLinesFiltersWithRegex []logline.LogLineFilterWithRegex,
-) error {
-	logTail, err := tail.TailFile(filepath, tail.Config{
-		Location: &tail.SeekInfo{
-			Offset: 0,
-			Whence: io.SeekEnd, // start tailing from end of log file
-		},
-		ReOpen:      false,
-		MustExist:   true,
-		Poll:        false,
-		Pipe:        false,
-		Follow:      true,
-		MaxLineSize: 0,
-		RateLimiter: nil,
-		Logger:      logrus.StandardLogger()})
-	if err != nil {
-		return stacktrace.Propagate(err, "An error occurred while attempting to tail the log file.")
+	conjunctiveLogLinesFiltersWithRegex []logline.LogLineFilterWithRegex) error {
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Debugf("Context was canceled, stopping streaming service logs for service '%v'", serviceUuid)
+			return nil
+		default:
+			jsonLogStr, err := getCompleteJsonLogString(logsReader)
+			if isValidJsonEnding(jsonLogStr) {
+				if err = strategy.sendJsonLogLine(jsonLogStr, logsByKurtosisUserServiceUuidChan, serviceUuid, conjunctiveLogLinesFiltersWithRegex); err != nil {
+					return err
+				}
+			}
+
+			if err != nil {
+				// if we've reached end of logs, return success, otherwise return the error
+				if errors.Is(err, io.EOF) {
+					return nil
+				} else {
+					return err
+				}
+			}
+		}
 	}
+}
 
-	for logLine := range logTail.Lines {
-		err = strategy.sendJsonLogLine(logLine.Text, logsByKurtosisUserServiceUuidChan, serviceUuid, conjunctiveLogLinesFiltersWithRegex)
-		if err != nil {
-			return stacktrace.Propagate(err, "An error occurred sending json log line '%v'.", logLine.Text)
+// tail -n X
+func (strategy *PerWeekStreamLogsStrategy) streamTailLogs(
+	ctx context.Context,
+	logsReader *bufio.Reader,
+	numLogLines uint32,
+	logsByKurtosisUserServiceUuidChan chan map[service.ServiceUUID][]logline.LogLine,
+	serviceUuid service.ServiceUUID,
+	conjunctiveLogLinesFiltersWithRegex []logline.LogLineFilterWithRegex) error {
+	tailLogLines := make([]string, 0, numLogLines)
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Debugf("Context was canceled, stopping streaming service logs for service '%v'", serviceUuid)
+			return nil
+		default:
+			jsonLogStr, err := getCompleteJsonLogString(logsReader)
+			if isValidJsonEnding(jsonLogStr) {
+				// collect all log lines in tail log lines
+				tailLogLines = append(tailLogLines, jsonLogStr)
+				if len(tailLogLines) > int(numLogLines) {
```

Comment:
I guess you are handling the empty string issue, in next PR. Make sense.

Reply:
it's not too explicit but actually, the `isValidJsonEnding` check should ensure that no empty log lines are streamed!